cmd/activator/main.go

```diff
@@ -133,6 +150,84 @@ func main() {
 		logger.Fatalw("Failed to create stats reporter", zap.Error(err))
 	}
 
+	// Set up signals so we handle the first shutdown signal gracefully.
+	stopCh := signals.SetupSignalHandler()
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, defaultResyncInterval)
+	servingInformerFactory := servinginformers.NewSharedInformerFactory(servingClient, defaultResyncInterval)
+	endpointInformer := kubeInformerFactory.Core().V1().Endpoints()
+	revisionInformer := servingInformerFactory.Serving().V1alpha1().Revisions()
+
+	go endpointInformer.Informer().Run(stopCh)
+	go revisionInformer.Informer().Run(stopCh)
+
+	params := queue.BreakerParams{QueueDepth: breakerQueueDepth, MaxConcurrency: breakerMaxConcurrency, InitialCapacity: 0}
+
+	// Return the number of endpoints, 0 if no endpoints are found.
+	endpointsGetter := func(revID activator.RevisionID) (int32, error) {
+		endpointKey := cache.ExplicitKey(revID.Namespace + "/" + reconciler.GetServingK8SServiceNameForObj(revID.Name))
+		endpoints, exists, err := endpointInformer.Informer().GetIndexer().Get(endpointKey)
+		if err != nil {
+			return 0, fmt.Errorf("Error getting endpoints for revision %v", revID)
+		}
+		if !exists {
+			// No endpoints exist yet.
+			return 0, nil
+		}
+		if exists && endpoints != nil {
+			endpoint := endpoints.(*corev1.Endpoints).Subsets
+			addresses := activator.EndpointsAddressCount(endpoint)
+			return int32(addresses), nil
+		}
+		return 0, nil
+	}
+
+	// Return the revision from the observer.
+	// Additionally to observer errors an error is returned if either no revision exists yet
+	// or we could not cast the interface to Revision for some reason.
+	revisionGetter := func(revID activator.RevisionID) (revision *v1alpha12.Revision, err error) {
+		revisionKey := cache.ExplicitKey(revID.Namespace + "/" + revID.Name)
+		rev, ok, err := revisionInformer.Informer().GetIndexer().Get(revisionKey)
```

Comment:
These two lines are equivalent to:

```golang
revision, err := revisionInformer.Lister().Revisions(revID.Namespace).Get(revID.Name)
```

Same above for endpoints. Don't be thrown off by this being called a Lister. It gives it a Lister interface but it's backed by the Indexer.

If you need to check for not-found, you can do so by importing `apierrors "k8s.io/apimachinery/pkg/api/errors"` and then do `if apierrors.IsNotFound(err) {`

Reply:
Got it, then we could delegate the existence checking to the lister. Nice to cut some lines.